[MKL-DNN] MKL-DNN specific Tensor modification

jczaja opened the pull request:
This is first of series of PRs on tensor modifications for MKL-DNN operations that are replacing mkl-dnn format storing with MKL-DNN primitive descriptor storing.

Idea is that currently MKL-DNN format is stored  in Tensor. format is not good enough for Transpose MKL-DNN as format is ambigious in terms of idenditifcations of actual mkldnn-layout. More over it does require recreation of primitive descriptions for given tensor. 

Adventages:
-  Design should be clearer
-  Redundant recreation of user primitive desc should be removed

In this PR:
- Tensor was modified to store MKLDNN primitive desc
-  Transpose op was adjusted to work with new functions
-  ConvGrad MKL-DNN was barely adjusted as well as Gaussian Random mkl-dnn.
-  All other mkl-dnn ops were not modified . They will be adjusted in next PRs (convolution will come next)
- **No changes to non MKL-DNN PaddlePaddle (core PaddlePaddle concepts are not modified)**

jczaja commented:
@panyx0718 To have CI passed I need your approve as PR is modifying tensor.h file. Coule you please take a look and comment/approve ?

luotao1 commented:
> Transpose op was adjusted to work with new functions

Do you have some benchmark of transpose op before and after this PR?

jczaja commented:
@luotao1 This changes do not improve performance. After all PRs are merged, performance should be a tiny bit better. But it is more redesign MKL-DNN tensor to make transpose MKL-DNN working and have cleaner design. I know you were working on c-API transfomer (diffrent flavour of transformer than python transformer: https://github.com/PaddlePaddle/Paddle/pull/14900) . How can I test this new C-API transformer (python transformer currently does not work due to: https://github.com/PaddlePaddle/Paddle/issues/14988) ?

jczaja commented:
@luotao1 This PR does not improve performance. It is on making Transpose MKL-DNN working
in every scenario & make foundation for cleaner design. Performance results were made using resnet50 to make sure performance did not go down. Actually with this PR it is a bit slower (x < 0.5%), but will be improved with next PR. I would like to make extensive performance tests (when I have other ops modified)

jczaja commented:
@luotao1 I updated PR with some changes to convolution , so performance on resnet-50 (SKX 8180, B1 & B128) is restored to original level.

jczaja commented:
@luotao1 @panyx0718 Do you need any more information to have this PR reviewed?

tensor-tang commented on paddle/fluid/framework/tensor.h:
```diff
@@ -37,23 +37,84 @@ class Tensor {
-
-  mkldnn::memory::format format_ = mkldnn::memory::format::format_undef;
+  mutable mkldnn::memory::primitive_desc mem_pd_;
 #endif
```
I think there is too much MKLDNN things in base `Tensor`.

MKLDNNTensor is better to read for adding library.

Need @panyx0718 approve this I think.

luotao1 replied:
@tensor-tang How do think about https://github.com/PaddlePaddle/Paddle/pull/11040#discussion_r192349103

tensor-tang replied:
That 's just how this current version comes from, but not the reason we cloud continue adding more.

`primitive_desc` contains more info like dims, format and data_type which is duplicated with many conceptions in tensor itself.

luotao1 commented:
@tensor-tang How do think about https://github.com/PaddlePaddle/Paddle/pull/11040#discussion_r192349103

tensor-tang commented on paddle/fluid/framework/tensor.h:
```diff
@@ -37,23 +37,84 @@ class Tensor {
 
-  inline void set_format(const mkldnn::memory::format format) {
-    format_ = format;
+  void CreateMemoryPrimDescFromDims(
```
1. please unify the name style. like `setMKLDNNMemPrimDesc`, `CreateMemoryPrimDescFromDims`
2. MemoryPrimDesc specify the  data_type, but tensor do not aware this only if get data.

jczaja replied:
1. It will be done
2. Yes. This is correct. data() / mutable_data() , returns pointer to allocation. Op has to ask separately for layout and/or Memory Primitive descriptor. Perhaps it would be good to  extend function data()/ mutable_data() so they can be aware of layouts and return converted data. Definitely having data() function that can return pointer to data in Paddle Layout , would help with implementing unit test so they can operate on block layout , and then we get result in paddle layout (NCHW or other) for comparison with reference. As paddlepaddle supports many devices (CPU,GPU,MKL-DNN...) I would like someone more familiar with other devices to trigger change in interface.

panyx0718 commented:
@sneaxiy Can you help take a look and see if there is a cleaner way?

panyx0718 commented on paddle/fluid/framework/tensor.h:
```diff
@@ -37,23 +37,84 @@ class Tensor {
+    }
+  }
+
+  // TODO(jczaja): This is depracted and will be removed
```
it will be removed? why add it?

jczaja replied:
@panyx0718 format() and set_format() where modified . And both they will be removed. Why they are present is that all(apart from transpose mkldnn op) MKL-DNN ops are using set_format() , format() functions. but mkl-dnn format is not used internally in data_layout_transform . So As We want to step by step replace all mkl-dnn ops ' set_format() with setMKLDNNMemPrimDesc(). After all mkl-dnn ops have set_format() removed and replaced, then We can remove set_format()/ format() implementation.

panyx0718 commented on paddle/fluid/framework/tensor.h:
```diff
@@ -37,23 +37,84 @@ class Tensor {
  public:
-  inline mkldnn::memory::format format() const { return format_; }
+  // TODO(jczaja): This is depracted and will be removed
+  inline mkldnn::memory::format format() const {
```
what if these format are put in Paddle's Layout enum? will it be cleaner without these ifdef PADDLE_WITH_MKLDNN

jczaja replied:
@panyx0718 Yes, but as we are to move away from format based conversions of Tensors, then there is no point of doing so. Moreover MKL-DNN format does "map" some MKL-DNN layout configuration eg. dims and strides etc. There are no enough format names so they map to all possible configurations of layouts. Transpose MKL-DNN op is using layouts that do not have format assigned to them , other than generic name: mkldnn_blocked. So it is not that format <=> mkl-dnn layout. format is not enough , to represent mkl-dnn layouts , so it has to go

panyx0718 commented on paddle/fluid/framework/tensor.h:
```diff
@@ -37,23 +37,84 @@ class Tensor {
    */
-
-  mkldnn::memory::format format_ = mkldnn::memory::format::format_undef;
+  mutable mkldnn::memory::primitive_desc mem_pd_;
```
what does this contain? Can it be mapped to current tensor apis? such as numel,layer?

jczaja replied:
@panyx0718 mem_pd_ is a primitive descriptor of MKL-DNN eg. This a structure that describes the implementation details of memory primitive eg. object does represent memory. So in that case mem_pd_ describes the dims of memory allocations , sizes, layout(way how memory is arranged, strides), and location of memory eg. cpu (MKL-DNN currently only supports CPU).

jczaja commented:
@tensor-tang ,
1. names of functions are unified.
2. I tried to create MKLDNNTensor as suggessted. Eg. Inheritance chain of  LODTensor  -> Tensor, now becomes LODTensor -> MKLDNNtensor -> Tensor .  This approach require quite alot of changes
In data_transform.cc , data_layout_layout_transform.cc  and in operator.cc (possibly more..) As Tensor is quite important entity tigtly integrated with ops and data loading. So it does require alot of changes eg. ifdef PADDLE_WITH_MKLDNN code.
    Other approach I tried is to Rename Tensor to TensorBase  and then for non-MKLDNN , alias TensorBase as Tensor , and for mkl-dnn builds MKLDNNTensor inherits from TensorBase and MKLDNNTensor is renamed to Tensor. This approach is also quite complex , aldo debugging it later may be complex.

So what are your suggestions on how to create MKLDNNTensor so it match current design ?

panyx0718 commented on paddle/fluid/framework/tensor.h:
```diff
@@ -37,23 +37,84 @@ class Tensor {
 
-  inline void set_format(const mkldnn::memory::format format) {
-    format_ = format;
+  void create_prim_desc_from_dims(
```
this doesn't need to be a method of tensor? Can this be independent function and create prim_desc elsewhere and then call set_prim_desc_?

jczaja replied:
@panyx0718 I moved body of create_prim_desc_from_dims into utils file eg. mkldnn_utils.h  and left thin wrapper that only calls moved out function. Is this good enough or you suggest to remove entirely create_prim_desc_from_dims() ?

tensor-tang commented:
> LODTensor -> MKLDNNtensor -> Tensor 

How about MKLDNNtensor -> LODTensor -> Tensor

jczaja commented:
@tensor-tang I haven't tried to implement MKLDNNTensor -> LoDTensor -> Tensor, as by looking into ops that are clearly checking if Tensor is LoDTensor etc. I suspected lots of changes . So I updated PR with original concept (no MKLDNNTensor) , but I moved most of MKLDNN functionality away to mkldnn_utils.h . Eg. Tensor contains thin wrapper functions to MKLDNN specific functions.

panyx0718 commented on paddle/fluid/framework/tensor.h:
```diff
@@ -37,23 +41,51 @@ class Tensor {
+    layout_ = DataLayout::kMKLDNN;
+  }
+
+  inline void create_prim_desc_from_dims(
```
I think the whole function can be moved to util? Then call set_mkldnn_prim_desc?

jczaja replied:
Ok, I moved whole functions to util and called set_mkldnn_prim_desc

panyx0718 commented on paddle/fluid/framework/tensor.h:
```diff
@@ -37,23 +41,51 @@ class Tensor {
 
-  inline void set_format(const mkldnn::memory::format format) {
-    format_ = format;
+  inline mkldnn::memory::primitive_desc get_prim_desc() const {
```
this method and the ones above should be called `mkldnn_prim_desc` to make its usage more obvious.

jczaja replied:
@panyx0718 Ok, methods renamed as suggested.

jczaja commented:
@panyx0718 , @tensor-tang , @luotao1  I updated review following your suggestions. Please review

panyx0718 commented on paddle/fluid/platform/mkldnn_utils.h:
```diff
@@ -0,0 +1,72 @@
+  auto md = mkldnn::memory::desc({ltz}, data_type, format);
+  auto& pool = platform::DeviceContextPool::Instance();
+  auto place = paddle::platform::CPUPlace();
+  auto* dev_ctx = dynamic_cast<platform::MKLDNNDeviceContext*>(pool.Get(place));
```
check nullptr?

jczaja replied:
@panyx0718 good suggestion

luotao1 commented on paddle/fluid/platform/mkldnn_utils.h:
```diff
@@ -0,0 +1,72 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
There is `paddle/fluid/platform/mkldnn_reuse.h`, why not combine `mkldnn_utils.h` into `mkldnn_reuse.h`?

tensor-tang replied:
Agreed.

jczaja replied:
@luotao1 , @tensor-tang  mkldnn_utils.h contains routines that are used within tensor.h. We cannot move mkldnn_utils.h into mkldnn_reuse.h or mkldnn_helper.h , as both of those headers are including tensor.h (via operator.h). So we would have tensor.h including mkldnn_reuse.h , mkldnn_reuse.h including tensor.h  ... which is cyclic dependency. That is why it is separate file

tensor-tang commented on paddle/fluid/operators/mkldnn/transpose_mkldnn_op.cc:
```diff
@@ -61,6 +61,17 @@ class TransposeMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
+    // so we need to create new primitive descriptor with changed logical layout
+    // so
+    // it match output shape
+
```
redundant line.

tensor-tang commented on paddle/fluid/operators/mkldnn/transpose_mkldnn_op.cc:
```diff
@@ -112,6 +124,17 @@ class TransposeMKLDNNGradOpKernel : public paddle::framework::OpKernel<T> {
+    // so we need to create new primitive descriptor with changed logical layout
+    // so
+    // it match output shape
+
```
redundant line.

tensor-tang commented on paddle/fluid/platform/mkldnn_utils.h:
```diff
@@ -0,0 +1,72 @@
+#include <mkldnn.h>
+#include <string>
+
+#ifdef PADDLE_WITH_MKLDNN
```
Why `mkldnn_helper.h` and `mkldnn_reuse.h` do not need this?

jczaja replied:
@tensor-tang Good catch. mkldnn_utils, should not have mentioned line either. This line was removed.

tensor-tang commented on paddle/fluid/platform/mkldnn_utils.h:
```diff
@@ -0,0 +1,72 @@
+#ifdef PADDLE_WITH_MKLDNN
+
+namespace paddle {
+
```
redundant line.

jczaja commented:
@panyx0718 , @luotao1 , @tensor-tang I adjusted changes to your suggestions. Please take a look. CI cannot pass without @panyx0718 approval.
